User username as VM username rather than random ID

marrobi opened the pull request:
Fixes #1148 #3693

Users email prior to the @ sign (as external users have unhelpful UPNs), or if no email UPN prior to the @ sign, as the username, limits to 20 chars.

Needed to update resource processor to pick OwnerId from the message and used AD TF provider to get details from user objectID.

![image](https://github.com/user-attachments/assets/ddbb5bca-f7c6-4c4e-9bdb-d9cd337ae9ff)

![image](https://github.com/user-attachments/assets/130c95a4-3f84-48b8-bdec-81859f0cd66e)

marrobi commented:
/test-extended

jonnyry commented:
@marrobi ok to update this branch?  will give it a test locally

marrobi commented:
> @marrobi ok to update this branch? will give it a test locally

Done.

jonnyry commented:
> > @marrobi ok to update this branch? will give it a test locally
> 
> Done.

Gave it a quick test on a local instance and worked nicely.

jonnyry reviewed (approved):
LGTM.  Couple of small comments re username calculation.

jonnyry commented:
@marrobi Re the username calculation logic -

```
 admin_username = (
    length(data.azuread_user.user.mail) > 0 ?
    substr(element(split("@", data.azuread_user.user.mail), 0), 0, 20) :
    substr(
      contains(element(split("@", data.azuread_user.user.user_principal_name), 0), "#EXT#") ?
      element(split("#EXT#", element(split("@", data.azuread_user.user.user_principal_name), 0)), 0) :
      element(split("@", data.azuread_user.user.user_principal_name), 0),
      0, 20
    )
  )
```

We use internal member accounts, and so would prefer if the UPN was used first - so the username aligns with the user's Entra username, rather than external email accounts which will come from a variety of domains and potentially in different formats.

Though understand a lot of people will be using external guest accounts.

Could the logic be something like this instead, so the UPN is preferred unless its a guest account:

```
if ((UPN contains "#EXT#") and (email is not blank)) then
  use email address
else
  use UPN
```

I think that would look something like:

```
 admin_username = (
    ((length(data.azuread_user.user.mail) > 0) && (contains(data.azuread_user.user.user_principal_name, "#EXT#"))) ?
    substr(element(split("@", data.azuread_user.user.mail), 0), 0, 20) :
    substr(
      contains(element(split("@", data.azuread_user.user.user_principal_name), 0), "#EXT#") ?
      element(split("#EXT#", element(split("@", data.azuread_user.user.user_principal_name), 0)), 0) :
      element(split("@", data.azuread_user.user.user_principal_name), 0),
      0, 20
    )
  )
```

jonnyry commented on templates/workspace_services/guacamole/user_resources/guacamole-azure-linuxvm/terraform/locals.tf:
```diff
@@ -7,12 +7,24 @@ locals {
+    length(data.azuread_user.user.mail) > 0 ?
+    substr(element(split("@", data.azuread_user.user.mail), 0), 0, 20) :
+    substr(
+      contains(element(split("@", data.azuread_user.user.user_principal_name), 0), "#EXT#") ?
```
Is there any value in doing the split if you are doing a contains check for "#EXT#"?

I.e. `contains(data.azuread_user.user.user_principal_name, "#EXT#")` is simpler

jonnyry commented on templates/workspace_services/guacamole/user_resources/guacamole-azure-windowsvm/terraform/locals.tf:
```diff
@@ -7,12 +7,24 @@ locals {
+    length(data.azuread_user.user.mail) > 0 ?
+    substr(element(split("@", data.azuread_user.user.mail), 0), 0, 20) :
+    substr(
+      contains(element(split("@", data.azuread_user.user.user_principal_name), 0), "#EXT#") ?
```
Is there any value in doing the split if you are doing a contains check for "#EXT#"?

I.e. `contains(data.azuread_user.user.user_principal_name, "#EXT#")` is simpler

jonnyry commented on templates/workspace_services/guacamole/user_resources/guacamole-azure-windowsvm/terraform/locals.tf:
```diff
@@ -7,12 +7,24 @@ locals {
-  vm_password_secret_name        = "${local.vm_name}-admin-credentials"
+  admin_username = (
+    length(data.azuread_user.user.mail) > 0 ?
+    substr(element(split("@", data.azuread_user.user.mail), 0), 0, 20) :
```
See https://github.com/microsoft/AzureTRE/pull/4333#issuecomment-2659169869 on preference of email over UPN

jonnyry commented on templates/workspace_services/guacamole/user_resources/guacamole-azure-linuxvm/terraform/locals.tf:
```diff
@@ -7,12 +7,24 @@ locals {
   storage_name                   = lower(replace("stg${substr(local.workspace_resource_name_suffix, -8, -1)}", "-", ""))
-  vm_password_secret_name        = "${local.vm_name}-admin-credentials"
+  admin_username = (
+    length(data.azuread_user.user.mail) > 0 ?
```
See https://github.com/microsoft/AzureTRE/pull/4333#issuecomment-2659169869 on preference of email over UPN

marrobi commented:
@jonnyry I've updated as per your PR comments, I've also added a  fix for the VM deletion issue #4135 

Are you able to give this another test? I don't have an environment deployed at the moment.

marrobi commented on templates/workspace_services/guacamole/user_resources/guacamole-azure-windowsvm/terraform/windowsvm.tf:
```diff
@@ -89,7 +79,7 @@ resource "azurerm_windows_virtual_machine" "windowsvm" {
   # (may be allowed once https://github.com/hashicorp/terraform-provider-azurerm/issues/25808 is fixed)
   #
-  lifecycle { ignore_changes = [tags, secure_boot_enabled, vtpm_enabled] }
+  lifecycle { ignore_changes = [tags, secure_boot_enabled, vtpm_enabled, admin_username] }
```
I believe Terraform will recreate the VM and the user would lose their work.

cc @jonnyry

jonnyry replied:
Yep -

[admin_username](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/windows_virtual_machine.html#admin_username-1) - (Required) The username of the local administrator used for the Virtual Machine. Changing this forces a new resource to be created.

Assume we don't want this to happen if people attempt to upgrade their existing VMs?

If they really want a machine with the new username format, they can delete their existing machine and create a new one.

marrobi commented on templates/workspace_services/guacamole/user_resources/guacamole-azure-windowsvm/terraform/locals.tf:
```diff
@@ -7,12 +7,19 @@ locals {
   storage_name                   = lower(replace("stg${substr(local.workspace_resource_name_suffix, -8, -1)}", "-", ""))
-  vm_password_secret_name        = "${local.vm_name}-admin-credentials"
+  admin_username = (
+    length(data.azuread_user.user.mail) > 0 && strcontains(data.azuread_user.user.user_principal_name, "#EXT#") ?
```
Ok, so if an email and is an external user, use before the @ on the email.

If not external, use the UPN, and if that contains #EXT#, use just before the #EXT#. If #EXT# does not exist I believe it will return the full string?

marrobi commented:
/test
